docs: advanced view features (structured projections and joins)

pvlugter opened the pull request:
Refs https://github.com/lightbend/kalix/issues/7323

Document some of the new advanced view features: structured projections and joins with multiple tables. Extend the customer registry example used in views to a simple store example with products and orders.

Adds an advanced view queries section, in a separate partial so it can be shared with Spring docs. Spring SDK will need some changes to support multiple tables, so will follow up with that separately.

These additional view features are not available by default. There's a note about this, and how to enable for local proxy and integration tests.

~Proxy version bump included here for now. Marking as draft until proxy release is finalised.~

pvlugter commented:
Should be ready now, with proxy 1.0.28.

pvlugter commented:
Realising that there are deeper changes needed for Views in the SDKs — given all the assumptions around a View having a single state type. So views with joins that don't transform work fine, but views that transform updates are not expressible in code right now — or more, only update transforms for the primary table would work as that's the single state type.

Switching the docs back to draft again, until some SDK changes are figured out, and for code generation and Spring SDK.

octonato commented:
> Realising that there are deeper changes needed for Views in the SDKs — given all the assumptions around a View having a single state type. So views with joins that don't transform work fine, but views that transform updates are not expressible in code right now — or more, only update transforms for the primary table would work as that's the single state type.
> 
> Switching the docs back to draft again, until some SDK changes are figured out, and for code generation and Spring SDK.

I just realised that this will be a problem for the Spring SDK. At some point we supported event handlers that would receive the current view state and the incoming event or value change. 

We reverted that option because we found it confusing and error prone. Also, we want it to be more aligned with the other components where we have a `currentState` method returning it. So, for views, we introduced a `viewState()` method. Well, that won't work anymore.

johanandren reviewed (commented):
Didn't review the sample in detail but rendered the docs locally and read through, the doc section looks good 👍

pvlugter commented:
Continuing ideas for supporting multi-table views with update transforms in https://github.com/lightbend/kalix/issues/7323#issuecomment-1321281453

pvlugter commented:
Updated samples to have customers and products be event sourced entities, with transform updates in the views, and orders to still be a value entity. Builds on the multi-table view support in #1299, rebased on that branch temporarily.

johanandren commented:
I mixed up this PR with #1299, so consider my review a review of both.

pvlugter commented on sdk/scala-sdk/src/main/scala/kalix/scalasdk/view/View.scala:
```diff
@@ -99,6 +99,6 @@ abstract class View[S] {
    *   is seen.
    */
-  def emptyState: S
+  def emptyState: S = null.asInstanceOf[S]
```
Makes sense. I'll restore in #1299. And update the comment, which mentions null.

pvlugter replied:
Restored in https://github.com/lightbend/kalix-jvm-sdk/pull/1299/commits/b4ccc3945b67bf895aad8279560282f9575e9660

pvlugter commented:
Integration tests for the views seem prone to timing out. Already a 20 second timeout on waiting for the view to update, but will increase a bit more.

pvlugter commented:
Can still fail with 60 second timeout. Only the Java version does this. Seems strange — is the different awaitility?

pvlugter commented:
Tracked down the integration test issue to a recently introduced dev proxy bug for in-memory journal. Fixed in https://github.com/lightbend/kalix-proxy/pull/1746.

pvlugter commented:
Bumped to proxy 1.0.29 for integration tests. No other changes since review.
